File: LeetCode/Easy/0108-convert-sorted-array-to-binary-search-tree/0108-convert-sorted-array-to-binary-search-tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public TreeNode SortedArrayToBST(int[] nums) {
        // using recurence as with N=10^4 balanced tree depth is log(N) ~< 14 and even if it raises, it's very limited rising
        return SortedArrayToBST(nums, 0, nums.Length);
    }

    private TreeNode SortedArrayToBST(int[] nums, int left, int right)
    {
        var mid = left + (right - left) / 2;

        return left < right
            ? new TreeNode(nums[mid], SortedArrayToBST(nums, left, mid), SortedArrayToBST(nums, mid + 1, right))
            : null;
    }
}

File: LeetCode/Easy/0108-convert-sorted-array-to-binary-search-tree/0108-convert-sorted-array-to-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
        left = 0
        right = len(nums)
        mid = (left + right) // 2
        root = TreeNode(val=nums[mid])

        stack: list[tuple[TreeNode, tuple[int, int], tuple[int, int]]] = [(root, (left, mid), (mid + 1, right))]

        while stack:
            node, (left_left, left_right), (right_left, right_right) = stack.pop()
            if right_left < right_right:
                mid = (right_left + right_right) // 2
                node.right = TreeNode(val=nums[mid])
                stack.append((node.right, (right_left, mid), (mid + 1, right_right)))
            
            if left_left < left_right:
                mid = (left_left + left_right) // 2
                node.left = TreeNode(val=nums[mid])
                stack.append((node.left, (left_left, mid), (mid + 1, left_right)))

        return root



        